src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -278,12 +293,25 @@ class CorrectUmis
     cachedResult match {
       case Some(result) => result
       case None         =>
-        val mismatches = umis.map(umi => Sequences.countMismatches(bases, umi))
-        val min        = mismatches.min
-        val matched    = (min <= maxMismatches) && (mismatches.count(m => m < min + this.minDistanceDiff) == 1)
-        val umiMatch   = UmiMatch(matched, umis(mismatches.indexOf(min)), min)
```

Comment:
changes below:
1. The while condition: when searching across UMIs, if we find a perfect match, and then we can break the loop if we find another match that is within the minimum distance
2. Pass the next-best distance found so far to the `countMismatches` method so it can abort early if possible
3.  Don't keep all the results, just the min and next-best, since that's all we need to create the `UmiMatch`